lib/puppet/defaults.rb

```diff
@@ -419,7 +419,20 @@ def self.default_diffargs
     :data_binding_terminus => {
       :type    => :terminus,
       :default => "hiera",
-      :desc    => "Where to retrieve information about data.",
+      :desc    =>
+        "This setting has been deprecated. Use of any value other than 'hiera' should instead be configured
+         in a version 5 hiera.yaml. Until this setting is removed, it controls which data binding terminus
+         to use for global automatic data binding (across all environments). By default this value is 'hiera'.
+         A value of 'none' turns off the global binding.",
+      :call_hook => :on_initialize_and_write,
+      :hook => proc do |value|
+        if Puppet[:strict] != :off
+          s_val = value.to_s # because sometimes the value is a symbol
+          unless s_val == 'hiera' || s_val == 'none' || value == '' || value.nil?
+            Puppet.deprecation_warning "Setting 'data_binding_terminus' is deprecated. Convert custom terminus to hiera 5 API."
```

Comment:
Actually, a custom terminus is not converted to hiera.yaml version 5 - the terminus has to be rewritten to be callable as a puppet function (implemented in Ruby or Puppet).
So - it is far more than just changing the configuration. We expect that there are close to 0 users that have written custom data binding termini, but we don't know for sure - hence this deprecation.

We have made a decision to using "hiera 5" as the name for "lookup". The hiera project codebase will remain at version 3 (versions 4 and 5 are inside the puppet code base).
We will continue to include hiera version 3 with puppet agent for backwards compatibility for backends until Puppet 6.0 or so when it will be removed. Thus giving hiera backend authors > 1 year to migrate to the hiera 5 API (a.k.a "lookup").

With all that I am saying that I think it is correct the way it is in the PR.